### PR TITLE
feat: support timeout for page.pdf() call

### PR DESCRIPTION
**Pull request by hottomato-c**
# Context
When use `page.pdf` API, there are some extreme scenarios that can cause the promise never get fulfilled or rejected. Hence this may cause potential problems. These kind of errors most likely happens in Chromium.

## Related Issues
- https://github.com/puppeteer/puppeteer/issues/6834
- https://github.com/puppeteer/puppeteer/issues/7496
- https://github.com/puppeteer/puppeteer/issues/5846

# Changes
- Added `timeout` parameter for PDFOption which will be used in `page.pdf` call

### CODE REVIEW

**Comment by hottomato-c** [comment]
> @googlebot I signed it!

@googlebot I signed it!

**Review by jschfflr — CHANGES_REQUESTED** [review]
Can you move to timeout to `createPDFStream` as it looks like the `Page.printToPDF` call there is the original source of the timeout?

**Comment by jschfflr** [comment]
Do you know if there's a corresponding bug reported for Chromium already? And if not, file one on https://crbug.com/new?
That way, we can make sure that we can also address the problem at its source. Thanks!

**Comment by hottomato-c** [comment]
> Can you move to timeout to `createPDFStream` as it looks like the `Page.printToPDF` call there is the original source of the timeout?

Sounds good ! Thanks for the review

**Comment by hottomato-c** [comment]
> Do you know if there's a corresponding bug reported for Chromium already? And if not, file one on https://crbug.com/new?
> That way, we can make sure that we can also address the problem at its source. Thanks!

np, will check on that .

**Comment by hottomato-c** [comment]
☑️  [Issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1248727) created for Chromium

**Comment by jschfflr** [comment]
Awesome, thanks! 
As we wanna make sure that the timeout still works when the bug is eventually fixed in Chromium, the tests should be updated to just use a timeout of `1` and make sure that the call times out. That's how we test the other time outs too. 
As for the test that you already wrote: It's a great reproduction that you can maybe put in a gist and reference it in the Chromium bug. That will help the engineer who'll be resolving the issue there a ton!

**Comment by hottomato-c** [comment]
> Awesome, thanks!
> As we wanna make sure that the timeout still works when the bug is eventually fixed in Chromium, the tests should be updated to just use a timeout of `1` and make sure that the call times out. That's how we test the other time outs too.
> As for the test that you already wrote: It's a great reproduction that you can maybe put in a gist and reference it in the Chromium bug. That will help the engineer who'll be resolving the issue there a ton!

Yep, the timeout values in tests are 1. 
[gist](https://gist.github.com/nikli2009/582e46afb11a1c3ca7cec6e503afa09b) has been created and linked to the Chromium issue :D 

Let me know if there are other things I could help with and thanks for taking time review my PR 😄

File: test/page.spec.ts [hunk 1]
```diff
@@ -1639,6 +1640,50 @@ describe('Page', function () {
...
...yRnSS7AQfcI3tC9orsGzkWOSU5B7kcuRlyB+T+gKB8ifwzBTYApapNfgAAAQAAAIUAOwAFAAAAAAACAAEAAgAWAAABAAFJAAAAAHjapZG9TgJBEMf/C4iChmBiDJVuic15QDCEwsSPSkMjJtR3eCCCd4Y7JFb3ChYWlsb4MPoSPoSVpf/dWwyJYuNNbue387E7MwtgUxQgoL75NaV1wmlNCWew9c1Z6m16RSbH3Q2kYcGoe8Mpeh4Mp+c4Q5nxEvJ4NJxlxrPhZZzjxfAKSvg0nENJFA3n0RRlw6vYF67hNfKT4SI2xLvhdWTFh+FX9j478w12qhC3JqEXyLbjh7Ju2/LM609Gzlh2PLcX+FGMIwTs8w5jDNDHJSL2XEYXO9RV2JQGyWWExAljQ+pDTBh9Cw8+cxxYtB1gRJFz54R651F71Cr6gpEWYv1XuNpaYrR4nooKmNHmeb7OHejV4R8x32G2h2vqMYa0Bej9WdEizyJ7m1Nw2LeqUuoalB7q2n1jnTIjYm+qtlPjsXSvEbOb2KVMtVi40ve4P+pKugh0ps+J/S/7t8nV9VQ7tLqckYpVb5rMu6X3Esf0dqlrtClfhfdXsafXhnn5CmpfaA13OgAAeNptz1cvg3EUx/HvoUMHWkrtvefzlC67Ouy9NwlaiSCVXrgnEuPWm3CH8vJo2v+lc/PJ75zknBzyyNZvGg//1QNInuSTjwEjJswUYMGKDTuFFFGMAycllOKijHLcVFBJFdXUUEsd9TTQSBPNtNBKG+100EkX3fTQSx/9aOiZ2wMM4sWHnwBBhhhmhFHGGGeCEJOEiRAlxhTTzDDLHPMssMgSy6ywyhrrbLDJFtvssMse+xxwyBHHnIiB+8wnbzyLkSdexSRmKRCLWMUmdimUIikWhzilREr54JNvfvgizSPv4pIyKedF3Kb45d1NQjenri40TYvkDGnKbPZkBkpd6VEOKAeVXqVP6VcGlEFlKKeu9uq69fwinkqenZ7cJnItTyynN2aIppLX2eCNhf8AwzpP+gB42tvB+L91A2Mvg/cGjoCIjYyMfZEb3di0IxQ3CER6bxAJAjIaImU3sGnHRDBsYFZw3cCs7bKBTcF1E7MDkzaYwwrksGlCOSxADqsslMMOUpYH4TBu4IBq5lRw3cXAUf+fgUl7I7NbmbZL5AYRbQDhkycvAAABUsGetAAA';
```
Comment (jschfflr) [inline test/page.spec.ts]: @nikli2009 Is this only happening for base64 decoded fonts or could this also be an url?

File: test/assets/pdf.html [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>PDF</title>
+    <style>
+      @font-face {
+        font-family: 'custom-font';
+        font-display: swap;
+        src: url(https://custom-font.com/font.woff2) format('woff2');
```
Comment (jschfflr) [inline test/assets/pdf.html]: Is it possible to use a local font here instead of loading something from a third party?

**Review by jschfflr — CHANGES_REQUESTED** [review]

**Comment by jschfflr** [comment]
@nikli2009 Thanks for updating the pull request! I left two more comments to reduce the size of the pr a bit - could you address them too?

File: test/assets/pdf.html (hunk 2 above)
Comment (hottomato-c) [reply test/assets/pdf.html]: that's a good point, will give it a go

File: test/page.spec.ts (hunk 1 above)
Comment (hottomato-c) [reply test/page.spec.ts]: 🤔 good question, gonna try some other combination / scenarios to see if it could break the pdf call as well

**Comment by hottomato-c** [comment]
> @nikli2009 Thanks for updating the pull request! I left two more comments to reduce the size of the pr a bit - could you address them too?

Happy to :D

**Comment by hottomato-c** [comment]
> @nikli2009 Thanks for updating the pull request! I left two more comments to reduce the size of the pr a bit - could you address them too?

@jschfflr  Thanks for the review :D, got some quick updates : 
1. After some manual tests, it seems that the printToPDF call only hangs when using base64(valid base64 tho) and external url at same time.
2. Removed those problematic headerTemplate config in the tests. I realized that as we set the `timeout` to be 1 ms and it will just always timeout (even when `printToPDF` being executed normally).

**Review by jschfflr — APPROVED** [review]
Ah yeah, right. Good catch!

**Pull request merged**